client/state/data-layer/extensions-middleware.js

```diff
@@ -1,101 +1,58 @@
 /**
- * External dependencies
- */
-import { toPairs, fromPairs } from 'lodash';
-
-/**
  * Internal dependencies
  */
-import { local, mergeHandlers } from './utils';
-
-const stripHandlersForKey = ( key, keyHandlers, keyHandlersToStrip ) => {
-	const newKeyHandlers = (
-		keyHandlersToStrip
-			? keyHandlers.filter( ( handler ) => ! keyHandlersToStrip.includes( handler ) )
-			: keyHandlers
+import { mergeHandlers } from './utils';
+import { middleware } from './wpcom-api-middleware';
+import debugFactory from 'debug';
+
+const debug = debugFactory( 'calypso:state:data-layer' );
+let extensionHandlers = Object.create( null );
+let handlerMiddleware = buildMiddleware( extensionHandlers );
+
+export function addHandlers( extensionName, handlers, existingHandlers = extensionHandlers ) {
+	if ( Object.keys( existingHandlers ).includes( extensionName ) ) {
+		throw new Error( `Handlers for extension ${ extensionName } already present.` );
+	}
+
+	// Set the module globals for convenience
+	extensionHandlers = { ...existingHandlers, [ extensionName ]: handlers };
+	handlerMiddleware = buildMiddleware( extensionHandlers );
+
+	return extensionHandlers;
+}
+
+export function removeHandlers( extensionName, existingHandlers = extensionHandlers ) {
+	const { [ extensionName ]: omitted, ...remainingHandlers } = existingHandlers;
+
+	if ( ! omitted ) {
+		debug( `Trying to remove nonexistent handlers for ${ extensionName }` );
+	}
+
+	// Set the module globals for convenience
+	extensionHandlers = remainingHandlers;
+	handlerMiddleware = buildMiddleware( extensionHandlers );
+
+	return extensionHandlers;
```

Comment:
there's a nice new symmetry to this pair of functions. I wonder if it would make sense to extract that…

```js
const state = {
	handlers: Object.create( null ),
	middleware: buildMiddleware( Object.create( null ) ),
};

/**
 * MUTATES INPUT!
 */
export const rebuildMiddleware = ( handlers, state = state ) => {
	state.handlers = handlers
	state.middleware = buildMiddleware( handlers );

	return handlers;
}

export const addHandlers = ( name, handlers, state = state ) => {
	// check for already-inserted

	return rebuildMiddleware( { ...state.handlers, [ name ]: handlers }, state ); 
}

export const removeHandlers = ( name, state = state ) => {
	const { [ name ]: omitted, ...handlers } = state.handlers;

	// check for not-already-there

	return rebuildMiddleware( handlers, state );
}
```

I don't like the way I've named and reused `state` but the idea comes through and I think this could be nice

 - takes advantage of the symmetry for code reuse
 - adds additional constraint on default `state` value
 - mutates input argument, something I usually like to avoid
 - direct mutation is an optimization, we could use `let` and return new state objects but we would have to `() => state` on every middleware call instead of directly accessing a variable. probably worth the optimization in this regard.